modules/darwinosl/darwinosl-source-oslog.m

```diff
@@ -0,0 +1,201 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "darwinosl-source-oslog.h"
+#include "messages.h"
+
+/* https://stackoverflow.com/questions/23684727/nsdateformatter-milliseconds-bug */
+static NSString *customBugFixDateFormat = @"yyyy-MM-dd'T'HH:mm:ss'.%06.0f'ZZZZZ";
+
+@implementation NSDateFormatter (BugFix)
+
+- (NSString *) stringFromDateWithMicroseconds:(NSDate *)theDate
+{
+  NSDateComponents *components = [self.calendar components:NSCalendarUnitNanosecond
+                                                fromDate:theDate];
+  double microseconds = lrint((double)components.nanosecond / (double)1000);
+  NSString *dateTimeFormatString = [self stringFromDate:theDate];
+  NSString *formattedString = [NSString stringWithFormat:dateTimeFormatString, microseconds];
+  return formattedString;
+}
+
+@end
+
+@interface OSLogSource ()
+
+@property (strong, nonatomic) OSLogStore *logStore;
+@property (strong, nonatomic) OSLogPosition *logPosition;
+@property (strong, nonatomic) OSLogEnumerator *enumerator;
+@property (assign, nonatomic) OSLogEnumeratorOptions enumeratorOptions;
+@property (strong, nonatomic) NSPredicate *filterPredicate;
+
+@end
+
+@implementation OSLogSource
+
++ (NSDateFormatter *) defaultDateFormatter
+{
+  NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
+  formatter.locale = [NSLocale localeWithLocaleIdentifier:@"en_US_POSIX"];
+  formatter.dateFormat = customBugFixDateFormat;
+  formatter.timeZone = [NSTimeZone localTimeZone];
+  return formatter;
+}
+
++ (NSDateFormatter *) RFC3339DateFormatter
+{
+  static NSDateFormatter *_RFC3339DateFormatter = nil;
+  if (_RFC3339DateFormatter == nil)
+    _RFC3339DateFormatter = OSLogSource.defaultDateFormatter;
+  return _RFC3339DateFormatter;
+}
+
+- (OSLogStore *) openStore
+{
+  g_assert(self.logStore == nil);
+  NSError *error = nil;
+
+  if (@available(macOS 12.0, *))
+    self.logStore = [OSLogStore storeWithScope:OSLogStoreSystem error:&error];
+  else
+    self.logStore = [OSLogStore localStoreAndReturnError:&error];
+
+  if (self.logStore == nil)
+    msg_error("darwinosl: Error getting handle to OSLog store",
+              evt_tag_long("error code", error.code),
+              evt_tag_str("error", error.localizedDescription.UTF8String));
+
+  return self.logStore;
+}
+
+- (OSLogEnumerator *) openEnumeratorWithDate:(NSDate *)startDate
```

Comment:
Can you rename this to either `openEnumeratorWithDateAndFilter` or just simply `openEnumerator`, I personally prefer the latter as based on the codeboth is optional, the startDate can be nil (for example if `self->options.do_not_use_bookmark == TRUE` ), and filterString also can be nil as well.

Reply:
This is part of the ObjectiveC naming convention, last tag in the function name is referring to the first function parameter, pls see the entriesEnumeratorWithOptions: function as a reference bellow